File: 0456-132-pattern/0456-132-pattern.py
class Solution:
    def find132pattern(self, nums):
        n = len(nums)
        if n < 3:
            return False

        min_left = [0] * n
        min_left[0] = nums[0]

        for i in range(1, n):
            min_left[i] = min(min_left[i - 1], nums[i])

        stack = []
        for j in range(n - 1, -1, -1):
            if nums[j] > min_left[j]:
                while stack and stack[-1] <= min_left[j]:
                    stack.pop()
                if stack and stack[-1] < nums[j]:
                    return True
                stack.append(nums[j])

        return False

File: LeetCode/0456-132-pattern/0456-132-pattern.java
import java.util.Stack;

class Solution {
    public boolean find132pattern(int[] nums) {
        int n = nums.length;
        if (n < 3) 
        {
            return false;
        }
        int[] minIndex = new int[n];
        minIndex[0] = nums[0];
        for (int i = 1; i < n; i++) 
        {
            minIndex[i] = Math.min(nums[i], minIndex[i - 1]);
        }
        Stack<Integer> s = new Stack<>();
        for (int j = n - 1; j >= 0; j--) 
        {
            if (nums[j] > minIndex[j]) 
            {
                while (!s.isEmpty() && s.peek() <= minIndex[j]) 
                {
                    s.pop();
                }
                if (!s.isEmpty() && s.peek() < nums[j]) 
                {
                    return true;
                }
            }
            s.push(nums[j]);
        }
        return false;
    }
}